rust-tvtf/src/funcs/filldown.rs

```diff
@@ -0,0 +1,416 @@
+            let scalars = params
+                .into_iter()
+                .filter(|p| p.is_scalar())
+                .collect::<Vec<_>>();
+
+            if !scalars.is_empty() {
+                let arg0 = scalars
+                    .first()
+                    .context("No column names parameter provided")?;
+
+                let Arg::String(column_names_str) = arg0 else {
+                    return Err(anyhow::anyhow!("Column names must be a string"));
+                };
+
+                // Parse comma-separated column names
+                column_names_str
+                    .split(',')
+                    .map(|s| s.trim().to_string())
+                    .filter(|s| !s.is_empty())
+                    .collect::<Vec<String>>()
+            } else {
+                Vec::new()
+            }
+        } else {
+            return Err(anyhow::anyhow!("Column names parameter is required"));
+        };
+
+        Ok(Filldown {
+            column_names,
+            last_values: Vec::new(),
+        })
+    }
+
+    fn ensure_last_value_len(&mut self, ncols: usize) {
+        if self.last_values.len() < ncols {
+            self.last_values.resize_with(ncols, || None);
+        }
+    }
+
+    fn need_filldown(&self, col_name: &str) -> bool {
+        if self.column_names.is_empty() {
+            return true;
+        }
+        for name in &self.column_names {
+            if name == col_name {
+                return true;
+            }
+            if name.ends_with('*') {
+                let prefix = name.trim_end_matches('*');
+                if col_name.starts_with(prefix) {
+                    return true;
+                }
+            }
+        }
+        false
+    }
+}
+
+fn extract_last_value(col: &ArrayRef) -> anyhow::Result<Option<ArrayRef>> {
+    let n = col.len();
+    if n == 0 {
+        return Ok(None);
+    }
+    for i in (0..n).rev() {
+        if col.is_valid(i) {
+            let idx_arr = Arc::new(arrow::array::UInt32Array::from(vec![i as u32])) as ArrayRef;
+            let single = take(col.as_ref(), idx_arr.as_ref(), None)
+                .context("Failed to extract last value for filldown")?;
+            return Ok(Some(single));
+        }
+    }
+    Ok(None)
+}
+
+impl TableFunction for Filldown {
+    fn process(&mut self, input: RecordBatch) -> anyhow::Result<Option<RecordBatch>> {
+        // create a new RecordBatch with same schema as input,
+        // replacing nulls with last non-null value for each column
+        let schema = input.schema();
+        let num_rows = input.num_rows();
+        let mut output_columns: Vec<ArrayRef> = Vec::with_capacity(input.num_columns());
+
+        self.ensure_last_value_len(input.num_columns());
+
+        for col_idx in 0..input.num_columns() {
+            let col = input.column(col_idx).clone();
+
+            if col.null_count() == 0 || !self.need_filldown(schema.field(col_idx).name()) {
```

Comment:
It seems not correct. If a column does not need filldown, in the current implementation, the code still calls extract_last_value.